### PR TITLE
[TF FE] Support Complex Tensors for Slice Operation

### PR DESCRIPTION
**Pull request by ysrastogi**
**Details:** Support Complex Tensors for Slice Operation

### CODE REVIEW

File: src/frontends/tensorflow_common/src/helper_transforms/flatten_tensor.hpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    // Create a vector from the data pointer.
+    std::vector<std::complex<float>> flat_tensor(data_ptr, data_ptr + num_elements);
+    
+    return flat_tensor;
+}
+
+}  // namespace op
+}  // namespace tensorflow
+}  // namespace frontend
+}  // namespace ov
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/helper_transforms/flatten_tensor.hpp:34]: please remove it. Not needed

File: src/frontends/tensorflow_common/src/op/slice.cpp [hunk 2]
```diff
@@ -24,35 +26,69 @@ namespace op {
...
     // assign stop values to the data shape
     Output<Node> stop_neg = make_shared<v3::ShapeOf>(input);
     stop_neg = make_shared<v1::ConvertLike>(stop_neg, size);
-
     // select the correct stop value based on a sign of size value
     auto negative_sizes_mask = make_shared<v1::Less>(size, const_zero);
     // TODO: investigate if we can simplify and replace Select with FloorMod operation
     // like FloorMod(size, input_shape)
     auto stop = make_shared<v1::Select>(negative_sizes_mask, stop_neg, stop_pos);
-
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/slice.cpp:48]: please revert this change

File: src/frontends/tensorflow_common/src/op/slice.cpp [hunk 3]
```diff
@@ -12,6 +13,7 @@
 #include "openvino/op/select.hpp"
 #include "openvino/op/shape_of.hpp"
 #include "utils.hpp"
+#include<complex>
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/slice.cpp:16]: include complex? unknown header, please remove

File: src/frontends/tensorflow_common/src/op/slice.cpp [hunk 4]
```diff
@@ -3,7 +3,8 @@
 //
 
 #include "openvino/op/slice.hpp"
-
+#include "helper_ops/flatten_tensor.hpp"
+#include "helper_ops/complex_tensor_reshape.hpp"
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/slice.cpp:7]: not needed, revert change

File: tests/layer_tests/tensorflow_tests/test_tf_Slice.py [hunk 5]
```diff
@@ -27,10 +27,24 @@ def create_slice_net(self, input_shape, input_type, begin_value, size_value):
...
                    ie_device, precision, ir_version, temp_dir=temp_dir,
                    use_legacy_frontend=use_legacy_frontend)
+    
+    @pytest.mark.parametrize("params", test_data_complex)
+    @pytest.mark.precommit_tf_fe
+    @pytest.mark.nightly
+    def test_slice_complex(self, params, ie_device, precision, ir_version, temp_dir, use_legacy_frontend):
+        self._test(*self.create_slice_net(**params),
+                   ie_device, precision, ir_version, temp_dir=temp_dir,
+                   use_legacy_frontend=use_legacy_frontend)
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_Slice.py:50]: not correctly implemented layer test.

**Review by rkazants — CHANGES_REQUESTED** [review]
please remove unneeded changes and double read instructions how to support complex type in GFI description.

File: src/frontends/tensorflow_common/src/helper_transforms/flatten_tensor.hpp (hunk 1 above)
Comment (ysrastogi) [reply src/frontends/tensorflow_common/src/helper_transforms/flatten_tensor.hpp]: Removed

File: src/frontends/tensorflow_common/src/op/slice.cpp (hunk 2 above)
Comment (ysrastogi) [reply src/frontends/tensorflow_common/src/op/slice.cpp]: Done

File: src/frontends/tensorflow_common/src/op/slice.cpp (hunk 4 above)
Comment (ysrastogi) [reply src/frontends/tensorflow_common/src/op/slice.cpp]: Done

File: src/frontends/tensorflow_common/src/op/slice.cpp (hunk 3 above)
Comment (ysrastogi) [reply src/frontends/tensorflow_common/src/op/slice.cpp]: Removed

File: tests/layer_tests/tensorflow_tests/test_tf_Slice.py (hunk 5 above)
Comment (ysrastogi) [reply tests/layer_tests/tensorflow_tests/test_tf_Slice.py:50]: How to implement layer test is there any guide?

**Comment by ysrastogi** [comment]
> please remove unneeded changes and double read instructions how to support complex type in GFI description.

I have removed the unneeded changes and modified slice.cpp according to GFI description.

File: tests/layer_tests/tensorflow_tests/test_tf_Slice.py (hunk 5 above)
Comment (ysrastogi) [reply tests/layer_tests/tensorflow_tests/test_tf_Slice.py]: Hii @rkazants I have implemented layer tests correctly

File: src/frontends/tensorflow_common/src/op/slice.cpp [hunk 6]
```diff
@@ -24,37 +26,54 @@ namespace op {
...
     Output<Node> stop_neg = make_shared<v3::ShapeOf>(input);
     stop_neg = make_shared<v1::ConvertLike>(stop_neg, size);
-
-    // select the correct stop value based on a sign of size value
     auto negative_sizes_mask = make_shared<v1::Less>(size, const_zero);
-    // TODO: investigate if we can simplify and replace Select with FloorMod operation
-    // like FloorMod(size, input_shape)
     auto stop = make_shared<v1::Select>(negative_sizes_mask, stop_neg, stop_pos);
-
-    // broadcast step value
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/slice.cpp:49]: please revert these unneeded changes

File: tests/layer_tests/tensorflow_tests/test_tf_Slice.py [hunk 7]
```diff
@@ -20,17 +31,45 @@ def create_slice_net(self, input_shape, input_type, begin_value, size_value):
...
-        self._test(*self.create_slice_net(**params),
+        self._test(self.create_slice_net(*params),
                    ie_device, precision, ir_version, temp_dir=temp_dir,
                    use_legacy_frontend=use_legacy_frontend)
+    
+
+    def test_slice_complex(self, params, ie_device, precision, ir_version, temp_dir, use_legacy_frontend):
+        self._test(self.create_complex_slice_net(*params),
+                   ie_device, precision, ir_version, temp_dir=temp_dir,
+                   use_legacy_frontend=use_legacy_frontend)
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_Slice.py:75]: please don't change existing layer tests and add separate test suite for complex type. See other PRs with examples of layer tests for complex type.

File: src/frontends/tensorflow_common/src/op/slice.cpp [hunk 8]
```diff
@@ -24,37 +26,54 @@ namespace op {
...
+
+    }
 
+    else{
     auto res = make_shared<v8::Slice>(input, start, stop, step);
     set_node_name(node.get_name(), res);
-    return res->outputs();
+    slice = res->outputs();
+    }
+    return slice;
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/slice.cpp:73]: please put under `if` only complex type case and leave non-complex type outside `if` after

File: src/frontends/tensorflow_common/src/op/slice.cpp (hunk 6 above)
Comment (ysrastogi) [reply src/frontends/tensorflow_common/src/op/slice.cpp]: I have revert all the unneeded changes

File: tests/layer_tests/tensorflow_tests/test_tf_Slice.py (hunk 7 above)
Comment (ysrastogi) [reply tests/layer_tests/tensorflow_tests/test_tf_Slice.py:81]: Done

File: src/frontends/tensorflow_common/src/op/slice.cpp [hunk 9]
```diff
@@ -24,10 +26,11 @@ namespace op {
 OutputVector translate_slice_op(const NodeContext& node) {
     default_op_checks(node, 3, {"Slice"});
     auto input = node.get_input(0);
+    auto complex_type_mark = as_type_ptr<ComplexTypeMark>(input.get_node_shared_ptr());
     auto start = node.get_input(1);
     auto size = node.get_input(2);
 
-    // create axiliary constants
+     // create axiliary constants
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/slice.cpp:33]:
```suggestion
    // create auxiliary constants
```

File: src/frontends/tensorflow_common/src/op/slice.cpp [hunk 10]
```diff
@@ -49,12 +52,37 @@ OutputVector translate_slice_op(const NodeContext& node) {
     // broadcast step value
     auto start_shape = make_shared<v3::ShapeOf>(start);
     auto step = make_shared<v3::Broadcast>(const_one, start_shape);
+    
+
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/slice.cpp:56]:
```suggestion
```

File: src/frontends/tensorflow_common/src/op/slice.cpp [hunk 11]
```diff
@@ -49,12 +52,36 @@ OutputVector translate_slice_op(const NodeContext& node) {
...
+        auto complex_slice = make_shared<ComplexTypeMark>(concat, complex_part_type);
+        return {complex_slice->output(0)};
+    }
+
 
     auto res = make_shared<v8::Slice>(input, start, stop, step);
     set_node_name(node.get_name(), res);
     return res->outputs();
+
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/slice.cpp:82]:
```suggestion
```

File: src/frontends/tensorflow_common/src/op/slice.cpp [hunk 12]
```diff
@@ -57,4 +83,4 @@ OutputVector translate_slice_op(const NodeContext& node) {
 }  // namespace op
 }  // namespace tensorflow
 }  // namespace frontend
-}  // namespace ov
+}  // namespace ov
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/slice.cpp:86]:
```suggestion
}  // namespace ov

```

File: src/frontends/tensorflow_common/src/op/slice.cpp [hunk 13]
```diff
@@ -24,10 +26,11 @@ namespace op {
 OutputVector translate_slice_op(const NodeContext& node) {
     default_op_checks(node, 3, {"Slice"});
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/slice.cpp:27]:
```suggestion
    default_op_checks(node, 3, {"Slice"}, true);
```

File: tests/layer_tests/tensorflow_tests/test_tf_Slice.py [hunk 14]
```diff
@@ -34,3 +33,50 @@ def test_slice_basic(self, params, ie_device, precision, ir_version, temp_dir, u
         self._test(*self.create_slice_net(**params),
                    ie_device, precision, ir_version, temp_dir=temp_dir,
                    use_legacy_frontend=use_legacy_frontend)
+    
+
+
+class TestComplexSlice(CommonTFLayerTest):
```
[the remark this answers is not in the document]
Comment (ysrastogi) [reply tests/layer_tests/tensorflow_tests/test_tf_Slice.py:39]: i have run these tests locally.

File: src/frontends/tensorflow_common/src/op/slice.cpp [hunk 15]
```diff
@@ -49,6 +52,29 @@ OutputVector translate_slice_op(const NodeContext& node) {
...
+        auto gather_index_real = make_shared<v0::Constant>(element::i32, Shape{}, 0);
+        auto gather_index_imag = make_shared<v0::Constant>(element::i32, Shape{}, 1);
+        auto minus_one = make_shared<v0::Constant>(element::i32, Shape{1}, -1);
+        auto real = make_shared<v8::Gather>(input, gather_index_real, minus_one)->output(0);
+        auto imag = make_shared<v8::Gather>(input, gather_index_imag, minus_one)->output(0);
+
+        auto real_part = make_shared<v8::Slice>(real,start,stop,step);
+        set_node_name(node.get_name(), real_part);
+        auto imag_part = make_shared<v8::Slice>(imag, start, stop, step);
+        set_node_name(node.get_name(), imag_part);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/slice.cpp:67]: please set node name for `concat` operation
